tests/conftest.py

```diff
@@ -223,9 +248,77 @@ def __init__(self, tmpdir, env):
         autostart.chmod(0o755)
         bin_path = os.path.join(BINDIR, 'herbstluftwm')
         self.proc = subprocess.Popen([bin_path, '--verbose'], env=env,
-                                stdout=subprocess.PIPE)
-        line = self.proc.stdout.readline()
-        assert line == b'hlwm started\n'
+                bufsize=0,  # essential for reading output with selectors!
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE)
+
+        sel = selectors.DefaultSelector()
+        sel.register(self.proc.stdout, selectors.EVENT_READ, data=sys.stdout)
+        sel.register(self.proc.stderr, selectors.EVENT_READ, data=sys.stderr)
+        self.output_selector = sel
+
+        # Wait for marker output from wrapper script:
+        self.read_and_echo_output(until_stdout='hlwm started')
+
+    def read_and_echo_output(self, until_stdout=None, until_stderr=None):
+        expect_sth = (until_stdout or until_stderr) is not None
+        max_wait = 30
+
+        stderr = ''
+        stdout = ''
+
+        def match_found():
+            if until_stdout and (until_stdout in stdout):
+                return True
+            if until_stderr and (until_stderr in stderr):
+                return True
+            return False
+
+        started = datetime.now()
+        while (datetime.now() - started).total_seconds() < max_wait:
+            select_timeout = 1
+            # If we're not polling for a matching string (anymore), there is no
+            # need for a dampening timeout:
+            if not expect_sth or match_found():
+                select_timeout = 0
+            selected = self.output_selector.select(timeout=select_timeout)
+            for key, events in selected:
+                # Read only single byte, otherwise we might block:
+                ch = key.fileobj.read(1).decode('ascii')
+
+                # Pass it through to the real stdout/stderr:
+                key.data.write(ch)
+                key.data.flush()
+
+                # Store in temporary buffer for string matching:
+                if key.fileobj == self.proc.stderr:
+                    stderr += ch
+                if key.fileobj == self.proc.stdout:
+                    stdout += ch
+
+            if selected != []:
```

Comment:
You can just do `if selected:`

Reply:
Somehow, I *knew* you we're going to say this. ;-)
I did this on purpose to make it more obvious that `selected` is either a non-empty list (implying that the `for` loop had something to iterate over) or it's empty. It's not `False`, it's not `None`, it's exactly `[]` *or* a non-empty list.

I'm aware that technically there is no difference, but I'm communicating to future contributors here, not to the interpreter. I don't want to make them look up the selectors API just because they wonder what else could be in `selected`.

Reply:
My $0.02: `if not selected:` clearly says "if nothing is selected". I've had to read `if selected != []:` a few times (because it feels like yoda reading a double negotiation or something) until I was sure I understood it the right way.